### DISCUSSION TITLE
Certificate based authentication

### DISCUSSION THREAD

**Discussion by inderpalpatheja** [strongswan/strongswan#1892] [General]
Hi,

I am running a usecase where i need that client only shall authenticate itself to the server by preseting its certificate and not the other way around. Is it possible that client do not request the server's certificate using any config.

I am using IKEv2.

Any help will be appreciated.

**Comment by tobiasbrunner** [comment]
> and not the other way around. Is it possible that client do not request the server's certificate using any config.

So how would you authenticate the server? What exactly is the use case here?

**Comment by inderpalpatheja** [comment]
Hi Tobias,

The use case here is that we have a client running on On premises hardware
and VPN server running in cloud...Client only has its own certificate which
it is presenting to server for the authentication. But we don't want to
keep one root CA certificate at the client as we don't want the server to
be authenticated by the client. Is it possible by any config..?

We have tried some configurations where we have put the send_cert variable
to never so that client does not request for server's certificate but still
authentication is failing and server is trying to look for its certificate
which the client has not even requested for.

Please provide some pointers if possible.

regards
Inder

regards
Inder

**Reply by tobiasbrunner** [reply]
> But we don't want to keep one root CA certificate at the client as we don't want the server to be authenticated by the client.

So anybody is allowed to impersonate the/a server and communicate with the client? Why bother with a VPN then?

>  Is it possible by any config..?

strongSwan doesn't support NULL authentication ([RFC 7619](https://datatracker.ietf.org/doc/html/rfc7619)). So both peers have to be authenticated in some way. If you don't want to use certificates, you could use pre-shared keys to authenticate the server (with the drawbacks PSKs have). Just configure the local and remote authentication accordingly. Make sure you use strong, random and especially client-specific PSKs. Otherwise, anybody (e.g. any other client) who knows the server's single PSK can again impersonate it. You still have to store the PSKs on the clients (and the server has to have access to all the client PSKs as well, so be aware of that risk, too).

**Comment by ventureElen** [comment]
Hi Tobias,

We don't want to use PSK at client side. client must authenticate using certificates but we can't keep root CAs at client side to authenticate server so can we use any other method of authentication at server other than certificate based? We tried setting leftauth=pubkey (for client) and leftauth=eap (for server) but it is giving error- "Non mutual EAP authentication is not allowed".  also we browsed through configuration examples we couldn't find any scenario where client auth mode is pubkey and server is using EAP or auth mode other than pubkey.  Is server auth with EAP in conjunction with client auth with pubkey not supported?

Is there any auth method by which server can authenticate to client without requiring to have a shared secret key or password configured at both client and server ends? 
Even if we succeeded in authenticating server using EAP method, it still requires secret key to be configured in ipsec.secrets or (secrets in swanctl.conf) at both ends. we don't want that.

**Reply by tobiasbrunner** [reply]
>  Is server auth with EAP in conjunction with client auth with pubkey not supported?

No, that's not possible by the protocol.

> Is there any auth method by which server can authenticate to client without requiring to have a shared secret key or password configured at both client and server ends?

Yes, certificates.

**Reply by ventureElen** [reply]
Can client authenticate server without keeping CA certificates? 
In the initiator response sent from server to client, it includes CA certificates it supports in CERTREQ payload so client should be able to authenticate server without requiring to have root CA certificates but when i removed CA certs from client, it failed to authenticate server with error something like "No valid RSA Public key found for..." so it looks like client is using CA certificates received from server to authenticate itself with server (by sending certificate issued by CAs supported by server in AUTH REQ) but it isn't using it to authenticate server. Please clarify.

Our usecase requires client to authenticate server without keeping CA certs.

**Reply by tobiasbrunner** [reply]
> Can client authenticate server without keeping CA certificates?

How would you imagine it do that? How would it trust the server certificate without root CA certificate?

> sent from server to client, it includes CA certificates it supports in CERTREQ payload so client should be able to authenticate server without requiring to have root CA certificates 

CERTREQ payloads don't contain certificates, just hashes to known CA certificates.

> Our usecase requires client to authenticate server without keeping CA certs.

Why?

**Reply by ventureElen** [reply]
> > CERTREQ payloads don't contain certificates, just hashes to known CA certificates.
>

Ok i thought server sends entire certificate so client can use that to authenticate server if it is sending hash then client also need root 
CAs.

> > Why?
>

Client will have a certificate issued by CA but it won't have access to root CA certificates. is it possible to authenticate server if its certificate is issued by root CA but client has intermediate CA certificate which is also issued by root CA but it doesn't have root CA certificate?

**Reply by tobiasbrunner** [reply]
The server does send its certificate (and intermediate  CA certificates). But it doesn't send the root CA certificate because how would the client trust it if it doesn't already have it?

> Client will have a certificate issued by CA but it won't have access to root CA certificates. 

That doesn't explain why. So why doesn't the client have access to the root CA?

> is it possible to authenticate server if its certificate is issued by root CA but client has intermediate CA certificate which is also issued by root CA but it doesn't have root CA certificate?

No, why would you think that helps? You need the issuer certificate (or the certificate itself) to trust it.
